### PR TITLE
Add vectorized plotting functionality to improve performance

### PR DESCRIPTION
**Pull request by MrDiver**
## Overview: What does this pull request change?

- added vectorized logic to `coords_to_points` in `Axes`
- added `use_vectorized: bool` parameter to `plot` in `CoordinateSystem`
- added `use_vectorized: bool` parameter to `ParametricFunction`
- added vectorized logic to `generate_points` in `ParametricFunction` depending on `self.use_vectorized`
- added `use_vectorized: bool` parameter to `plot_antiderivative_graph`
- added vectorized logic to `plot_antiderivative_graph`
- added vectorized logic to `number_to_point` in `NumberLine`
- vectorized `inverse_function` in `scale.py`
- vectorized `angle_of_vector` in `space_ops.py`

- changed naming of test data in `frames_comparison.py` to support fixture parameters because ":" is not allowed in a file name
- added `use_vectorized` pytest fixture
- added `use_vectorized` fixture to
    - `test_axes::test_plot_functions`
    - `test_axes::test_plot_derivative_graph`
    - `test_axes::test_plot`
    - `test_axes::test_get_riemann_rectangles`
- added test `test_coordinate_systems::test_plot_log_x_axis_vectorized`
- added test `test_coordinate_system::test_point_to_coords_vectorized`
- added test `test_coordinate_system::test_coords_to_point_vectorized`
- added test `test_space_ops::test_angle_of_vector_vectorized`
- added test `test_number_line::test_number_to_point`
- added test `test_number_line::test_point_to_number`
- added test `test_space_ops::test_angle_of_vector_vectorized`

Todo for Future PR
- [ ] CoordinateSystem.plot_implicit_curve
- [ ] CoordinateSystem.plot_polar_graph
- [ ] ImplicitFunction
- [ ] CoordinateSystem.pr2pt

## Motivation and Explanation: Why and how do your changes improve the library?

Manim does a ton of function calls internally and reducing them step by step might be a good idea. At some point we could make a breaking change and vectorize functions by default to remove branching code versions. The performance increase is not that much but a step in the right direction. I think this vectorized version could also improve VectorField or other Classes that use `coords_to_points` 🤔

| Before | After Non-Vectorized | Vectorized |
|-|-|-|
|grafik|grafik|grafik|
|grafik `coords_to_point` (yellow) on the right taking 1/3 of `generate_points`|grafik| grafik `coords_to_point` on the left taking a lot less|

### Table Calls/Percentage
Function | Before | After Non-Vectorized | Vectorized |
|-|-|-|-|
`number_to_point` | 327 / 6.3 | 407 / 6.5% | 58 / 1.0% |
`coords_to_point` | 486 / 8% | 517 / 8.2% | 68 / 1.3% |
`generate_points` | 1481 / 24.3% | 1508 / 24.0% | 978 / 18.8% |
**Time (s)** | $42.33$ | $43.48$ | $36.39$ |

As we can see the implementations of `number_to_point` and `coords_to_point` get a little bit slower but i tried to reduce the overhead as much as possible in the non vectorized case to not slow down the library overall. But when more and more functions adapt to the vectorized style this will be neglible because we can see that `generate_points` overall gets faster.

Data collected via `vprof` and performance difference is from 42 to 36 seconds.
To view the [profile_data.zip](https://github.com/ManimCommunity/manim/files/8788980/profile_data.zip) use 
```bash
# original main branch
vprof -i reference.json
# this PR
vprof -i vectorized.json
vprof -i not_vectorized.json
```

# Scene used for testing
Bench_ManimCE_v0 15 2

```py
from manim import *
class Bench(Scene):
    def construct(self):
        ax1 = NumberPlane((-3, 3), (-4, 4))
        track = ValueTracker(0)
        def func(t):
            out =  np.array([np.sin(2 * (t+track.get_value())), np.sin(3 * (t+track.get_value())), 0])
            return out

        vectorized = False
        graph = ax1.plot_parametric_curve(func, color=RED, t_range=np.array([0, 1]), use_vectorized=vectorized)
        graph2 = ax1.plot(lambda x: x**2, use_vectorized=vectorized)
        graph3 = ax1.plot_derivative_graph(graph2, use_vectorized=vectorized)

        graph.add_updater(lambda m: m.become(ax1.plot_parametric_curve(func, color=RED, t_range=np.array([0, 1]), use_vectorized=vectorized)))
        graph2.add_updater(lambda m: m.become(ax1.plot(lambda x: (x**3)*track.get_value()-x/2, use_vectorized=vectorized)))
        graph3.add_updater(lambda m: m.become(ax1.plot_derivative_graph(graph2, use_vectorized=vectorized)))
        rects = ax1.get_riemann_rectangles(graph)
        rects.add_updater(lambda m: m.become(ax1.get_riemann_rectangles(graph)))

        self.add(ax1, graph, graph2, graph3, rects)
        self.play(track.animate.set_value(TAU),run_time=10)
```

## Further Information and Comments

This is a extended implementation of
- #2513 

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by MrDiver** [comment]
Maybe it would be a good idea to add a `use_vectorized` option in the config dictionary

**Comment by behackl** [comment]
> Maybe it would be a good idea to add a `use_vectorized` option in the config dictionary. This would remove a lot of the unnecessary parameter passing and would make extending this to other functions easier!

I'm not sure about this. The respective classes or objects should hold this sort of behavioral information directly, the global config should only do this for classes that the user should normally not access (like the renderer, the camera, the file writer, ...). And what if you'd want to use vectorized functions in some parts, but not in others?

I'm open to further discussion, but there has to be a very clear and thought through proposal if we want to head in this direction. :-)

**Comment by MrDiver** [comment]
> And what if you'd want to use vectorized functions in some parts, but not in others?
This is not really possible anyway, beacuse then the function all need to contain logic to decide what their input is and if they need to translate an input array to to elements and then go through a for loop and apply itself. The idea behind that is to slowly optimize the code behind the scenes and give the option to use it but not result in breaking changes so that some function are not working correctly anymore.

The only thing that would still be useful is on the high level end of the plot functions to decide manually if someone would like to have `f(x)` or `f([x_0,x_1,...])`

One option would also be to completely vectorize the low level functions in any way or add a vectorized version for all of them. But this results in a lot of code doubling. And when the goal is to optimize in the long term it seems rather counter productive.

File: manim/mobject/graphing/coordinate_systems.py [hunk 1]
```diff
@@ -1354,6 +1362,8 @@ def plot_antiderivative_graph(
             The y-value at which the graph intercepts the y-axis.
         samples
             The number of points to take the area under the graph.
+        use_vectorized
+            Whether to use the vectorized version of the antiderivative.
```
Comment (kilacoda) [inline manim/mobject/graphing/coordinate_systems.py:1366]: Maybe a small one line description of what vectorised means here, or a reference to `plot` could be helpful for a first time user of this function to know what exactly this parameter does?

File: manim/mobject/graphing/coordinate_systems.py [hunk 2]
```diff
@@ -1829,21 +1839,48 @@ def _create_axis(
         axis.shift(-axis.number_to_point(self._origin_shift([axis.x_min, axis.x_max])))
         return axis
 
-    def coords_to_point(self, *coords: Sequence[float]) -> np.ndarray:
+    def coords_to_point(self, *coords: Sequence[float] | np.ndarray) -> np.ndarray:
```
Comment (kilacoda) [inline manim/mobject/graphing/coordinate_systems.py:1842]: According to the docstring that follows this, shouldn't there also be a `Sequence[Sequence[float]]` for better typing? Something of the sort:
```suggestion
    def coords_to_point(self, *coords: Sequence[float] | Sequence[Sequence[float]] | np.ndarray) -> np.ndarray:
```

Also, not sure about this, but wouldn't `np.ndarray` already be covered as a `Sequence` here, and could thus be removed?

File: manim/mobject/graphing/number_line.py [hunk 3]
```diff
@@ -324,24 +324,40 @@ def get_tick_range(self) -> np.ndarray:
 
         return self.scaling.function(tick_range)
 
-    def number_to_point(self, number: float) -> np.ndarray:
+    def number_to_point(self, number: float | np.ndarray) -> np.ndarray:
```
Comment (kilacoda) [inline manim/mobject/graphing/number_line.py:327]: Not necessarily an `np.ndarray`, as shown by the example below this in the docstring, where a regular list is used.
```suggestion
    def number_to_point(self, number: float | Sequence[float]) -> np.ndarray:
```

File: manim/mobject/graphing/number_line.py [hunk 4]
```diff
@@ -357,13 +373,27 @@ def point_to_number(self, point: Sequence[float]) -> float:
...
+
         """
+        point = np.asarray(point)
         start, end = self.get_start_and_end()
         unit_vect = normalize(end - start)
         proportion = np.dot(point - start, unit_vect) / np.dot(end - start, unit_vect)
         return interpolate(self.x_min, self.x_max, proportion)
 
-    def n2p(self, number: float) -> np.ndarray:
+    def n2p(self, number: float | np.ndarray) -> np.ndarray:
```
Comment (kilacoda) [inline manim/mobject/graphing/number_line.py:396]: See comment for `number_to_point`

File: manim/mobject/graphing/coordinate_systems.py (hunk 2 above)
Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py:1842]: Yes, Sequence probably includes a `np.ndaray` but the problem is that a `Sequence` does not require the implementation of mathematical operators which are heavily used in the vectorized versions. So i wanted to make explicit that you can either pass a sequence containing floats or a corresponding ndarray which implements all the needed functionality.

File: manim/mobject/graphing/number_line.py (hunk 3 above)
Comment (MrDiver) [reply manim/mobject/graphing/number_line.py:327]: See comment above

File: manim/mobject/graphing/coordinate_systems.py (hunk 1 above)
Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py]: Done

File: manim/mobject/graphing/coordinate_systems.py (hunk 2 above)
Comment (kilacoda) [reply manim/mobject/graphing/coordinate_systems.py:1844]: Got it, but the suggestion still holds valid right? The `Sequence[Sequence[float]]` one.

Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py:1844]: A single point is `Sequence[float]` a list of points is `np.ndarray`. `Sequence[Sequence[float]]` would just duplicate `np.ndarray` but without the needed functionality 🤔 .

Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py:1844]: i would be fine with changing it to `npt.NDarray[np.float64]` but that's why theres a typing discussion. We can probably delay the "correct" typing of this function.

Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py:1844]: Wait nvm ... i used `np.asarray`... youre right i haven't looked at my own code for too long! YES YES!!! 🥳

**Review by kilacoda — APPROVED** [review]
LGTM

File: manim/mobject/graphing/coordinate_systems.py [hunk 5]
```diff
@@ -639,6 +640,9 @@ def plot(
             The function used to construct the :class:`~.ParametricFunction`.
         x_range
             The range of the curve along the axes. ``x_range = [x_min, x_max, x_step]``.
+        use_vectorized
+            Whether to pass in the generated t value array to the function. Only use this if your function supports it.
+            Output should be a numpy array of shape ``[y_0,y_1,...]``
```
Comment (behackl) [inline manim/mobject/graphing/coordinate_systems.py:645]:
```suggestion
            Output should be a numpy array of shape ``[y_0, y_1, ...]``
```

File: manim/mobject/graphing/coordinate_systems.py [hunk 6]
```diff
@@ -1354,6 +1362,10 @@ def plot_antiderivative_graph(
             The y-value at which the graph intercepts the y-axis.
         samples
             The number of points to take the area under the graph.
+        use_vectorized
+            Whether to use the vectorized version of the antiderivative. This means
+            to pass in the generated t value array to the function. Only use this if your function supports it.
+            Output should be a numpy array of shape ``[y_0,y_1,...]``
```
Comment (behackl) [inline manim/mobject/graphing/coordinate_systems.py:1368]:
```suggestion
            Output should be a numpy array of shape ``[y_0, y_1, ...]``
```

File: manim/mobject/graphing/coordinate_systems.py [hunk 7]
```diff
@@ -1829,21 +1841,50 @@ def _create_axis(
...
+            np.around(ax.coords_to_point([[0, 1], [1, 1], [1, 0]]), 2)
+            # array([[0.  , 0.75, 0.  ],
+            #        [0.86, 0.75, 0.  ],
+            #        [0.86, 0.  , 0.  ]])
+            np.around(
+                ax.coords_to_point([0, 1, 1], [1, 1, 0]), 2
+            )  # Transposed version of the above
+            # array([[0.  , 0.86, 0.86],
+            #        [0.75, 0.75, 0.  ],
+            #        [0.  , 0.  , 0.  ]])
```
Comment (behackl) [inline manim/mobject/graphing/coordinate_systems.py:1886]: Is there a particular reason why you decided against writing this as a doctest?

```suggestion
        .. code-block:: python

            >>> from manim import Axes
            >>> import numpy as np
            >>> ax = Axes()
            >>> np.around(ax.coords_to_point(1, 0, 0), 2)
            array([0.86, 0.  , 0.  ])
            >>> np.around(ax.coords_to_point([[0, 1], [1, 1], [1, 0]]), 2)
            array([[0.  , 0.75, 0.  ],
                   [0.86, 0.75, 0.  ],
                   [0.86, 0.  , 0.  ]])
            >>> np.around(
            ...    ax.coords_to_point([0, 1, 1], [1, 1, 0]), 2
            ... )  # Transposed version of the above
            array([[0.  , 0.86, 0.86],
                   [0.75, 0.75, 0.  ],
                   [0.  , 0.  , 0.  ]])
```

File: manim/mobject/graphing/coordinate_systems.py [hunk 8]
```diff
@@ -1865,26 +1906,69 @@ def construct(self):
...
+
+            from manim import Axes, RIGHT
+            import numpy as np
+
+            ax = Axes(x_range=[0, 10, 2])
+            np.around(ax.point_to_coords(RIGHT), 2)
+            # array([5.83, 0.  ])
+            np.around(ax.point_to_coords([[0, 0, 1], [1, 0, 0]]), 2)
+            # array([[5.  , 0.  ],
+            #       [5.83, 0.  ]])
```
Comment (behackl) [inline manim/mobject/graphing/coordinate_systems.py:1969]:
```suggestion
        .. code-block:: python

            >>> from manim import Axes, RIGHT
            >>> import numpy as np
            >>> ax = Axes(x_range=[0, 10, 2])
            >>> np.around(ax.point_to_coords(RIGHT), 2)
            array([5.83, 0.  ])
            >>> np.around(ax.point_to_coords([[0, 0, 1], [1, 0, 0]]), 2)
            array([[5.  , 0.  ],
                  [5.83, 0.  ]])
```

File: manim/mobject/graphing/functions.py [hunk 9]
```diff
@@ -31,6 +31,9 @@ class ParametricFunction(VMobject, metaclass=ConvertToOpenGL):
     use_smoothing
         Whether to interpolate between the points of the function after they have been created.
         (Will have odd behaviour with a low number of points)
+    use_vectorized
+        Whether to pass in the generated t value array to the function as ``[t_0,t_1,...]``. Only use this if your function supports it.
+        Output should be a numpy array of shape ``[[x_0,x_1,...],[y_0,y_1,...],[z_0,z_1,...]]`` but z can also be 0 if the Axes is 2D
```
Comment (behackl) [inline manim/mobject/graphing/functions.py:36]: Lines are a bit long, breaking:

```suggestion
        Whether to pass in the generated t value array to the function as ``[t_0, t_1, ...]``.
        Only use this if your function supports it. Output should be a numpy array
        of shape ``[[x_0, x_1, ...], [y_0, y_1, ...], [z_0, z_1, ...]]`` but ``z`` can
        also be 0 if the Axes is 2D
```

**Review by behackl — APPROVED** [review]
Thank you very much for this! I have no further comments regarding the code, just some minor remarks regarding documentation. :-)

File: manim/mobject/graphing/coordinate_systems.py (hunk 7 above)
Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py:1886]: it gives different results on different machines 🤷🏼 i suspected it was the pixel scaling thing with virtual size but im not sure... but it gives the same result when doing frames_comparison so ... i really don't know

File: manim/mobject/graphing/coordinate_systems.py (hunk 8 above)
Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py:1969]: same here as above

**Pull request merged**